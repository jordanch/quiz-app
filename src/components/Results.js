import React from "react"
import classNames from "classnames"
import Typography from "@material-ui/core/Typography"
import { withStyles } from "@material-ui/core/styles"
import Button from "@material-ui/core/Button"
import PropTypes from "prop-types"
import { Link } from "react-router-dom"
import Done from "@material-ui/icons/Done"
import Clear from "@material-ui/icons/Clear"
import Card from '@material-ui/core/Card';
import { isUserCorrect } from "../util/quiz_entity"

const styles = (theme) => ({
	container: {
		width: "100%",
		height: "100%",
		backgroundColor: "white",
		display: "flex",
		flexDirection: "column",
		alignItems: "center",
		justifyContent: "space-around"
	},
	score: {
		borderBottom: "1px solid black",
		width: "100%"
	},
	results: {
		overflowY: "scroll"
	},
	resultRow: {
    margin: "20px 0",
		padding: "10px 0",
		"&:first-child": {
			marginTop: "20px"
		}
	},
	resultRowAnswer: {
		display: "flex",
		alignItems: "center",
		margin: "0 0 15px 0"
	},
	resultRowAnswerDetails: {
		display: "flex",
		alignItems: "center",
		justifyContent: "space-around",
		fontSize: "12px"
	},
	resultRowItemIcon: {
		flexBasis: "10%"
  },
  resultsRowItemIconCorrect: {
    color: "green"
  },
  resultsRowItemIconIncorrect: {
    color: "red"
  },
	resultRowItemText: {
    flexBasis: "90%",
    padding: "0 20px"
	},
	button: {
		margin: theme.spacing.unit
	},
	buttonLink: {
		textDecoration: "none"
	},
	icon: {
		margin: theme.spacing.unit,
		fontSize: 32
	}
})

const Results = (props) => {
	const { classes, quizEntities } = props

	return (
		<div className={classes.container}>
			<div className={classes.score}>
				<Typography variant="headline" gutterBottom align="center">
					You scored
				</Typography>

				<Typography variant="headline" gutterBottom align="center">
					{`${quizEntities.reduce(
						(acc, curr) => acc + (isUserCorrect(curr) ? 1 : 0),
						0
					)} of ${quizEntities.length}`}
				</Typography>
			</div>

			<div className={classes.results}>
				{quizEntities.map((quizEntity) => {
					const isUserAnswerCorrect = isUserCorrect(quizEntity)
					return (
						<Card
							className={classes.resultRow}
							key={quizEntity.appId}
						>
							<div className={classes.resultRowAnswer}>
								{quizEntity.userAnswer === quizEntity.correctAnswer && (
									<Done className={classNames(classes.resultRowItemIcon, classes.resultsRowItemIconCorrect)}/>
								)}

								{quizEntity.userAnswer !== quizEntity.correctAnswer && (
									<Clear className={classNames(classes.resultRowItemIcon, classes.resultsRowItemIconIncorrect)} />
								)}

								<div className={classes.resultRowItemText}>
									<Typography
										variant="body2"
										align="left"
										className={
											isUserAnswerCorrect
												? classes.resultsRowCorrect
												: classes.resultsRowIncorrect
										}
									>
										{quizEntity.question}
									</Typography>
								</div>
							</div>

							<div className={classes.resultRowAnswerDetails}>
								<div>Your answer: {quizEntity.userAnswer.toString()}</div>
								<div>Correct answer: {quizEntity.correctAnswer.toString()}</div>
							</div>
						</Card>
					)
				})}
			</div>

			<Link to="/" className={classes.buttonLink}>
				<Button variant="contained" color="primary" className={classes.button}>
					PLAY AGAIN?
				</Button>
			</Link>
		</div>
	)
}

Results.propTypes = {
	classes: PropTypes.object.isRequired,
	quizEntities: PropTypes.array.isRequired
}

export default withStyles(styles)(Results)
