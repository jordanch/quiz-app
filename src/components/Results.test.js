import React from "react"
import Results from "./Results"
import renderer from "react-test-renderer"
import { MemoryRouter } from "react-router"


it("renders 10 quiz entities correct with some right and some wrong", () => {
  const quizEntities = [
    {
      category: "General Knowledge",
      question:
        "This is the correct spelling of &quot;Supercalifragilisticexpialidocious&quot;.",
      appId: 1,
      userAnswer: true,
      correctAnswer: true
    },
    {
      category: "History",
      question:
        "Joseph Stalin&#039;s real name was Ioseb Bessarionis dze Dzugashvili.",
      appId: 2,
      userAnswer: true,
      correctAnswer: true
    },
    {
      category: "Entertainment: Video Games",
      question:
        "In &quot;Portal 2&quot;, Cave Johnson started out Aperture Science as a shower curtain company.",
      appId: 3,
      userAnswer: false,
      correctAnswer: true
    },
    {
      category: "Science: Mathematics",
      question:
        "If you could fold a piece of paper in half 50 times, its&#039; thickness will be 3/4th the distance from the Earth to the Sun.",
      appId: 4,
      userAnswer: false,
      correctAnswer: true
    },
    {
      category: "Entertainment: Film",
      question:
        "&quot;Cube&quot;, &quot;Cube 2: Hypercube&quot; and &quot;Cube Zero&quot; were directed by the same person.",
      appId: 5,
      userAnswer: false,
      correctAnswer: true
    },
    {
      category: "Entertainment: Books",
      question: "Harry Potter was born on July 31st, 1980.",
      appId: 6,
      userAnswer: true,
      correctAnswer: true
    },
    {
      category: "Entertainment: Television",
      question: "The Klingon home planet is Qo&#039;noS.",
      appId: 7,
      userAnswer: true,
      correctAnswer: true
    },
    {
      category: "Science: Mathematics",
      question:
        "L&#039;H&ocirc;pital was the mathematician who created the homonymous rule that uses derivatives to evaluate limits with indeterminations.",
      appId: 8,
      userAnswer: true,
      correctAnswer: false
    },
    {
      category: "Entertainment: Music",
      question:
        "Pete Townshend&#039;s solo album, &quot;White City: A Novel&quot;, is set in the metropolitan area of Chicago.",
      appId: 9,
      userAnswer: false,
      correctAnswer: false
    },
    {
      category: "General Knowledge",
      question:
        "Spoon theory is a theory, utilizing &quot;Spoons&quot; as a metaphor for energy they can use in a day.",
      appId: 10,
      userAnswer: true,
      correctAnswer: true
    }
  ]
  const tree = renderer
    .create(
      <MemoryRouter>
        <Results quizEntities={quizEntities} />
      </MemoryRouter>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
