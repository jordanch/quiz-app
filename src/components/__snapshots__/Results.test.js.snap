// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 10 quiz entities correct with some right and some wrong 1`] = `
<div
  className="Results-container-1"
>
  <div
    className="Results-score-2"
  >
    <h1
      className="MuiTypography-root-14 MuiTypography-headline-19 MuiTypography-gutterBottom-31 MuiTypography-alignCenter-27"
    >
      You scored
    </h1>
    <h1
      className="MuiTypography-root-14 MuiTypography-headline-19 MuiTypography-gutterBottom-31 MuiTypography-alignCenter-27"
    >
      6 of 10
    </h1>
  </div>
  <div
    className="Results-results-3"
  >
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconCorrect-8"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            This is the correct spelling of &quot;Supercalifragilisticexpialidocious&quot;.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          true
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconCorrect-8"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            Joseph Stalin&#039;s real name was Ioseb Bessarionis dze Dzugashvili.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          true
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconIncorrect-9"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            In &quot;Portal 2&quot;, Cave Johnson started out Aperture Science as a shower curtain company.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          false
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconIncorrect-9"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            If you could fold a piece of paper in half 50 times, its&#039; thickness will be 3/4th the distance from the Earth to the Sun.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          false
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconIncorrect-9"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            &quot;Cube&quot;, &quot;Cube 2: Hypercube&quot; and &quot;Cube Zero&quot; were directed by the same person.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          false
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconCorrect-8"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            Harry Potter was born on July 31st, 1980.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          true
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconCorrect-8"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            The Klingon home planet is Qo&#039;noS.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          true
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconIncorrect-9"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            L&#039;H&ocirc;pital was the mathematician who created the homonymous rule that uses derivatives to evaluate limits with indeterminations.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          true
        </div>
        <div>
          Correct answer: 
          false
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconCorrect-8"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            Pete Townshend&#039;s solo album, &quot;White City: A Novel&quot;, is set in the metropolitan area of Chicago.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          false
        </div>
        <div>
          Correct answer: 
          false
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root-40 MuiPaper-elevation1-43 MuiPaper-rounded-41 MuiCard-root-39 Results-resultRow-4"
    >
      <div
        className="Results-resultRowAnswer-5"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-67 Results-resultRowItemIcon-7 Results-resultsRowItemIconCorrect-8"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
        <div
          className="Results-resultRowItemText-10"
        >
          <aside
            className="MuiTypography-root-14 MuiTypography-body2-22 MuiTypography-alignLeft-26"
          >
            Spoon theory is a theory, utilizing &quot;Spoons&quot; as a metaphor for energy they can use in a day.
          </aside>
        </div>
      </div>
      <div
        className="Results-resultRowAnswerDetails-6"
      >
        <div>
          Your answer: 
          true
        </div>
        <div>
          Correct answer: 
          true
        </div>
      </div>
    </div>
  </div>
  <a
    className="Results-buttonLink-12"
    href="/"
    onClick={[Function]}
  >
    <button
      className="MuiButtonBase-root-100 MuiButton-root-74 MuiButton-contained-85 MuiButton-containedPrimary-86 MuiButton-raised-88 MuiButton-raisedPrimary-89 Results-button-11"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-75"
      >
        PLAY AGAIN?
      </span>
      <span
        className="MuiTouchRipple-root-103"
      />
    </button>
  </a>
</div>
`;
